metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/RestFunctions.java

```diff
@@ -168,241 +195,399 @@ public Object apply(List<Object> args, Context context) throws ParseException {
+    HttpClientContext httpClientContext = HttpClientContext.create();
+    boolean credentialsAdded = false;
+    CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+
+    // Add the basic auth credentials if the rest config settings are present
+    if (restConfig.getBasicAuthUser() != null && restConfig.getBasicAuthPasswordPath() != null) {
+      String password = new String(readBytes(new Path(restConfig.getBasicAuthPasswordPath())), StandardCharsets.UTF_8);
+      credentialsProvider.setCredentials(
+              new AuthScope(target),
+              new UsernamePasswordCredentials(restConfig.getBasicAuthUser(), password));
+      credentialsAdded = true;
+    }
+
+    // Add the proxy basic auth credentials if the rest config settings are present
+    if (proxy.isPresent() && restConfig.getProxyBasicAuthUser() != null &&
+            restConfig.getProxyBasicAuthPasswordPath() != null) {
+      String password = new String(readBytes(new Path(restConfig.getProxyBasicAuthPasswordPath())), StandardCharsets.UTF_8);
+      credentialsProvider.setCredentials(
+              new AuthScope(proxy.get()),
+              new UsernamePasswordCredentials(restConfig.getProxyBasicAuthUser(), password));
+      credentialsAdded = true;
+    }
+    if (credentialsAdded) {
+      httpClientContext.setCredentialsProvider(credentialsProvider);
+    }
+    return httpClientContext;
+  }
+
+  /**
+   * Read bytes from a HDFS path.
+   * @param inPath
+   * @return
+   * @throws IOException
+   */
+  private static byte[] readBytes(Path inPath) throws IOException {
+    FileSystem fs = FileSystem.get(inPath.toUri(), new Configuration());
+    try (FSDataInputStream inputStream = fs.open(inPath)) {
+      return IOUtils.toByteArray(inputStream);
+    }
+  }
+
+  /**
+   * Perform the HttpClient request and handle the results.  A configurable list of status codes are accepted and the
+   * response content (expected to be json) is parsed into a Map.  Values returned on errors and when response content
+   * is also configurable.  The rest config "timeout" setting is imposed in this method and will abort the get request
+   * if exceeded.
+   *
+   * @param restConfig
+   * @param httpRequestBase
+   * @return
+   * @throws IOException
+   */
+  protected static Object executeRequest(RestConfig restConfig, HttpRequestBase httpRequestBase) throws IOException {
+    URI uri = httpRequestBase.getURI();
+    HttpHost target = new HttpHost(uri.getHost(), uri.getPort(), uri.getScheme());
+    Optional<HttpHost> proxy = getProxy(restConfig);
+    HttpClientContext httpClientContext = getHttpClientContext(restConfig, target, proxy);
+    httpRequestBase.setConfig(getRequestConfig(restConfig, proxy));
+
+    // Schedule a command to abort the request if the timeout is exceeded
+    ScheduledFuture scheduledFuture = scheduledExecutorService.schedule(httpRequestBase::abort, restConfig.getTimeout(), TimeUnit.MILLISECONDS);
```

Comment:
Why do we need to manage timeouts on our end? e.g. https://www.baeldung.com/httpclient-timeout

Reply:
From the link you reference:

> HttpClient doesn’t have any configuration that allows us to set an overall timeout for a request; it does, however, provide abort functionality for requests, so we can leverage that mechanism to implement a simple timeout mechanism: